database/database.go

```diff
@@ -0,0 +1,160 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+var (
+	// dbName defines the name of the database this service uses
+	dbName = "pinner"
+	// collSkylinks defines the name of the collection which will hold
+	// information about skylinks
+	collSkylinks = "skylinks"
+
+	// mongoCompressors defines the compressors we are going to use for the
+	// connection to MongoDB
+	mongoCompressors = "zstd,zlib,snappy"
+	// mongoReadPreference defines the DB's read preference. The options are:
+	// primary, primaryPreferred, secondary, secondaryPreferred, nearest.
+	// See https://docs.mongodb.com/manual/core/read-preference/
+	mongoReadPreference = "primary"
+	// mongoWriteConcern describes the level of acknowledgment requested from
+	// MongoDB.
+	mongoWriteConcern = "majority"
+	// mongoWriteConcernTimeout specifies a time limit, in milliseconds, for
+	// the write concern to be satisfied.
+	mongoWriteConcernTimeout = "30000"
+)
+
+type (
+	// DB holds a connection to the database, as well as helpful shortcuts to
+	// collections and utilities.
+	DB struct {
+		staticCtx    context.Context
+		staticDB     *mongo.Database
+		staticLogger *logrus.Logger
+	}
+
+	// DBCredentials is a helper struct that binds together all values needed for
+	// establishing a DB connection.
+	DBCredentials struct {
+		User     string
+		Password string
+		Host     string
+		Port     string
+	}
+)
+
+// New creates a new database connection.
+func New(ctx context.Context, creds DBCredentials, logger *logrus.Logger) (*DB, error) {
+	return NewCustomDB(ctx, dbName, creds, logger)
+}
+
+// NewCustomDB creates a new database connection to a database with a custom name.
+func NewCustomDB(ctx context.Context, dbName string, creds DBCredentials, logger *logrus.Logger) (*DB, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid context provided")
+	}
+	if logger == nil {
```

Comment:
here we error out while in other cases we init to a discard logger, I'm in favour of defaulting to a discard logger everywhere I think that's kinda nice

Reply:
I'm very much against defaulting to a discard logger.
This is a big decision about the way the service reports its state and it should be made consciously and not by default. The only place where I would use a discard logger is during testing. Otherwise we're flying blind in production.